### ISSUE TITLE
slow training

### ISSUE THREAD

**Issue by zhang-wen** [clab/dynet#156]
hello, every one,

i create a MT model with following configurations:

```
        self.trainer = AdadeltaTrainer(self.model, eps=1e-7, rho=0.99)

        self.layers = kwargs.pop('layer_cnt', 1)
        self.semb_dim = kwargs.pop('semb_dim', 50)
        self.temb_dim = kwargs.pop('temb_dim', 50)
        self.enc_nhids = kwargs.pop('enc_nhids', 128)
        self.dec_nhids = kwargs.pop('dec_nhids', 128)
        self.att_dim = kwargs.pop('att_dim', 50)
        self.svoc_size = kwargs.pop('src_vocab_size', 30000)
        self.tvoc_size = kwargs.pop('trg_vocab_size', 30000)
        self.drop_rate = kwargs['dropout']

        self.enc_fwd_rnn = GRUBuilder(self.layers, self.semb_dim, self.enc_nhids, self.model)
        self.enc_bck_rnn = GRUBuilder(self.layers, self.semb_dim, self.enc_nhids, self.model)

        self.y2a_rnn = SimpleRNNBuilder(self.layers, self.temb_dim, self.dec_nhids, self.model)
        self.a2o_rnn = SimpleRNNBuilder(self.layers, 2 * self.enc_nhids, self.dec_nhids, self.model)
```

user parameters are:

```
		 source lookup_table => (30000, 50)
		 target lookup_table => (30000, 50)
		 attention_w1        => (50, 256)
		 attention_w2        => (50, 128)
		 attention_v         => (1, 50)
		 no_voc_w            => (30000, 434)
		 no_voc_b            => (30000,)
```

totally 29 weights parameters, i update once for each batch with batch size 80, 

```
batch_persent_loss = network.get_batch_loss(bx, bxm, by, bym)
batch_persent_loss.backward()
network.trainer.update()
```

but i test the time, the result is as following, it take too much time, i think.

```
forward       10.26s 
backward    17.64s 
update         2.06s

```
are there something wrong with my training ? why is it so slow ?
it perplex me for a long time .
any reply will be appreciated very much ...

**Comment by neubig** [comment]
Sorry, we're going to need more info. Why do you think this is taking too long? For example, if this is much slower than other MT toolkits with the same parameters, then please send the commands that you used to run both of the toolkits and we can start to know if this is a bug or not.

**Comment by zhang-wen** [comment]
okay, sorry about that, actually i am not very sure, let me think about this and try to construct some examples

**Comment by zhang-wen** [comment]
i use theano before, but i think dynet is easier to debug than theano, so now i want to use dynet.

for same parameters:
i run my MT system by theano command:

` THEANO_FLAGS=mode=FAST_RUN,device=gpu0,floatX=float32,lib.cnmem=5000, python main.py`

the forward and update time for one batch (batch size is 80) is about: 
`Update time[3.806s] `

and by dynet command:

`python main.py --dynet-gpu --dynet_mem 8000 --dynet-gpu-ids 1`

the forward and backward time for updating one batch (same batch size) is about:
```
54.594s
[upd/s=>54.594s/80=0.682s]
```

**Comment by neubig** [comment]
Yes, if you use the lookup_batch function it will be much much faster.

**Comment by zhang-wen** [comment]
hi @neubig  is following network a large network ? 
it spends about 20s for updating one batch (80 sentences).
i am not sure whether it is normal .
i heard you have a large model to train, how about your speed, can i make a comparison ?

```
init network parameters done:
    src_emb_lookuptable : (30000, 256)
    trg_emb_lookuptable : (30000, 256)
    GRU_Enc_fwd_1_W_z   : (512, 768)
    GRU_Enc_fwd_1_W_r   : (512, 768)
    GRU_Enc_fwd_1_W_h   : (512, 768)
    GRU_Enc_fwd_1_h0    : 512
    GRU_Enc_bwd_2_W_z   : (512, 768)
    GRU_Enc_bwd_2_W_r   : (512, 768)
    GRU_Enc_bwd_2_W_h   : (512, 768)
    GRU_Enc_bwd_2_h0    : 512
    GRU_Dec_Wz_1        : (512, 768)
    GRU_Dec_Wr_1        : (512, 768)
    GRU_Dec_Wh_1        : (512, 768)
    attention_W_1       : (128, 1024)
    attention_W_2       : (128, 512)
    attention_v         : (1, 128)
    GRU_Dec_Wz_2        : (512, 1536)
    GRU_Dec_Wr_2        : (512, 1536)
    GRU_Dec_Wh_2        : (512, 1536)
    GRU_Dec_h0_2        : 512
    combine_out_W       : (256, 1792)
    combine_out_b       : 256
    logistic_W          : (30000, 256)
    logistic_b          : 30000
```

**Comment by neubig** [comment]
Sorry about the late reply. Looking at the size of this net, I think I've gotten over 1000 words/sec on GPU on nets that are slightly smaller than this, so your speeds definitely sound too low. If you're using CPU, then maybe the speeds are reasonable.

**Comment by neubig** [comment]
Recently there have been a number of speed improvements to DyNet (particularly on GPU with mini-batches). If slow training is still an issue, I'd suggest trying the newest version, and if you're still having problems there is probably something up with the implementation.

**Comment by zhang-wen** [comment]
@neubig  hi , you are welcome, i will try again and test the training speed, let you know,  thank you very much.

**Comment by zhang-wen** [comment]
@neubig  hi, neubig, can you show me your training code ? i try the newest version, but it is the same , very slow, may be my code have some problems, because i use mask, what do you mean by 1000 words/sec, it includes the time of forward , backward and update calculation ?

**Comment by neubig** [comment]
Sure, here are some reference implementations of machine translation implemented in DyNet:
https://github.com/clab/dynet/blob/master/examples/python/attention.py
https://github.com/neubig/lamtram
https://github.com/odashi/nmtkit

**Comment by neubig** [comment]
Perhaps this is resolved? If so I'll close it.

**Comment by zhang-wen** [comment]
ok, may be my implemetion has some problem, let me check again. thank you @neubig

**Comment by zhang-wen** [comment]
i write a GRU myself, and i want to test how long it will take to calculate one time, the forward function is as following:

```
        z_i = logistic(self.W_z_xh * x_i + self.U_z_xh * self.h_im1 + self.b_z_xh)

        r_i = logistic(self.W_r_xh * x_i + self.U_r_xh * self.h_im1 + self.b_r_xh)

        h_i_below = self.activation(self.W_h_xh * x_i +
                                    self.U_h_xh * cmult(r_i, self.h_im1) +
                                    self.b_h_xh
                                   )

        h_i = cmult((1 - z_i), self.h_im1) + cmult(z_i, h_i_below)

        if x_m is not None:
            batch_size = x_m.shape[0]
            m_matrix = np.reshape(np.tile(x_m, (1, self.output_dims)), \
                                  (batch_size * self.output_dims))
            x_m_expr = reshape(inputVector(m_matrix), (self.output_dims,), batch_size)
            _1m_x_m_expr = reshape(inputVector(1.-m_matrix), (self.output_dims,), batch_size)
            h_i = cmult(x_m_expr, h_i) + cmult(_1m_x_m_expr, self.h_im1)
```

the weight matrix size is: 

```
    GRU_Enc_fwd_1_W_z   : (512, 256)
    GRU_Enc_fwd_1_U_z   : (512, 512)
    GRU_Enc_fwd_1_b_z   : (512,)
    GRU_Enc_fwd_1_W_r   : (512, 256)
    GRU_Enc_fwd_1_U_r   : (512, 512)
    GRU_Enc_fwd_1_b_r   : (512,)
    GRU_Enc_fwd_1_W_h   : (512, 256)
    GRU_Enc_fwd_1_U_h   : (512, 512)
    GRU_Enc_fwd_1_b_h   : (512,)
```

the time consuming of one forward calculation of this GRU is about 0.02s, is it normal ?

**Comment by neubig** [comment]
I think this is probably a problem of GRU calculation as noted here: https://github.com/clab/dynet/issues/351

I'll close this for the more descriptive one.
